CDAP-17981: Introduce FileFetcherInternalHttpHandler in AppFabric

wyzhang opened the pull request:
* FileFetcherInternalHttpHandler will be used to stream file content from appfab to other services. 
* This will be used in the next PR for file localizer to download files from appfab

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,108 @@
+    Location location = Locations.getLocationFromAbsolutePath(locationFactory, path);
+
+    if (!location.exists()) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Path %s not found", path));
```
Better throw a `NotFoundException` and let the exception handler to deal with the response.

wyzhang replied:
Done. Fixed.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,108 @@
+  public void download(HttpRequest request, HttpResponder responder) throws Exception {
+    String prefix = String.format("%s/location/", Constants.Gateway.INTERNAL_API_VERSION_3);
+    String path = request.uri().substring(prefix.length());
+    Location location = Locations.getLocationFromAbsolutePath(locationFactory, path);
```
I think we shouldn't allow fetching any location, as the client and workaround permission check to get access to artifact jar that it is not supposed to.

wyzhang replied:
Plan to add an access token validation. Caller provides a path along with an access token for this specific path. 

The token can be a short lived 5min access token produced by token service that allows access to a specific file. Appfab hands out the token along with the location path to other services, which call back to appfab to download the file.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,108 @@
+  @GET
+  @Path("/location/**")
+  public void download(HttpRequest request, HttpResponder responder) throws Exception {
+    String prefix = String.format("%s/location/", Constants.Gateway.INTERNAL_API_VERSION_3);
```
Should we do an enforcer check via the internal auth mechanism?

wyzhang replied:
If "internal auth mechanism" refers to the work of internal system identity auth, then yes.

chtyim commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java:
```diff
@@ -0,0 +1,108 @@
+    byte[] buf = new byte[64 * 1024];
+    InputStream inputStream = location.getInputStream();
+    try {
+      responder.sendContent(HttpResponseStatus.OK, new BodyProducer() {
```
Use the `LocationBodyProducer` class to send `Location` content.

wyzhang replied:
Done.

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/FileFetcherHttpHandlerInternalTest.java:
```diff
@@ -0,0 +1,163 @@
+
+  @BeforeClass
+  public static void setup() throws Exception {
+    Injector injector = Guice.createInjector(
```
Create a cConf and set the Constants. CFG_LOCAL_DATA_DIR To a folder created by the TEMP_FOLDER. Then add ConfigModule with the cConf to create the injector

wyzhang replied:
Done.

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/FileFetcherHttpHandlerInternalTest.java:
```diff
@@ -0,0 +1,162 @@
+    Location dst = new LocalLocationFactory().create(targetFile.toURI());
+
+    HttpResponse httpResponse = download(srcFile, dst);
+    Assert.assertEquals(httpResponse.getResponseCode(), HttpResponseStatus.NOT_FOUND.code());
```
The expected value should be the first parameter

wyzhang replied:
Done.

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/FileFetcherHttpHandlerInternalTest.java:
```diff
@@ -0,0 +1,162 @@
+
+    // Download the file.
+    File targetFile = new File(TEMP_FOLDER.newFolder(), "dst_file");
+    Location dst = new LocalLocationFactory().create(targetFile.toURI());
```
Use Locations.toLocation

wyzhang replied:
Done

chtyim commented on cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/FileFetcherHttpHandlerInternalTest.java:
```diff
@@ -0,0 +1,162 @@
+
+    // Download the file.
+    File targetFile = new File(TEMP_FOLDER.newFolder(), "dst_file");
+    Location dst = new LocalLocationFactory().create(targetFile.toURI());
```
Use Locations.toLocation

wyzhang replied:
Done.

chtyim reviewed (approved):
Couple comments left about the test class

wyzhang committed:
CDAP-17981: Introduce FileFetcherInternalHttpHandler in AppFabric

This is intended to be used for appfab to proxy files from distributed
file system, since it is possible that other services may not have
direct access to the distributed file system (e.g. in k8s where we
may want to run user code in isolated pods which don't have permission
to access distributed file system that appfab has access to)
